usr/sbin/p11sak/p11sak.c

```diff
@@ -1141,6 +1191,200 @@ static CK_ATTRIBUTE_TYPE col2type(int col)
+                }
+            }
+        }
+    }
+
+    printf(" 0 ");
+    return rc;
+}
+
+/**
+ *  Print in p11sak_defined_attrs.conf defined attributes in long format
+ */
+static CK_RV long_print_vendor(CK_SESSION_HANDLE session, CK_OBJECT_HANDLE hkey) 
+{
+    CK_RV rc = CKR_OK;
+    int f;
+    FILE *fp;
+    struct ConfigBaseNode *cfg, *c, *name, *hex_string, *type;
+    struct ConfigStructNode *structnode;
+    char file_loc[1024] = P11SAK_DEFINED_ATTRS_LOCATION;
+
+    // get file location
+    cfg = NULL;
+    if (strcmp(P11SAK_DEFINED_ATTRS_LOCATION, "") == 1) {
+        strcpy(file_loc, P11SAK_DEFINED_ATTRS_LOCATION);
+    } else {
+        strcpy(file_loc, P11SAK_DEFAULT_CONF_FILE);
+    } 
+
+    // open and parse file
+    if ((fp = fopen(file_loc, "r")) == NULL) {
+        fprintf(stderr, "Failed to load %s: %s\n", file_loc, strerror(errno));
+        return CKR_ARGUMENTS_BAD;
+    }
+    if (parse_configlib_file(fp, &cfg, error_hook, 0)) {
+        fprintf(stderr, "Failed to parse %s\n", file_loc);
+        return CKR_DATA_INVALID;
+    }
+
+    // parse and print attributes
+    fclose(fp);
+    if (cfg != NULL)
+    {
+        confignode_foreach(c, cfg, f) {
+            if (confignode_hastype(c, CT_STRUCT) && strcmp(c->key, "attribute") == 0) {
+                // parse...
+                structnode = confignode_to_struct(c);
+                name = confignode_find(structnode->value, "name");
+                hex_string = confignode_find(structnode->value, "id");
+                type = confignode_find(structnode->value, "type");
+
+                // checking syntax of attribute...
+                if (name == NULL || !confignode_hastype(name, CT_BAREVAL)) {
+                    fprintf(stderr, "Invalid name in Attribute in line %hu\n", c->line);
+                    return CKR_DATA_INVALID;
+                }
+                if (hex_string == NULL || !confignode_hastype(hex_string, CT_INTVAL)) {
+                    fprintf(stderr, "Invalid value in Attribute in line %hu\n", c->line);
+                    return CKR_DATA_INVALID;
+                }
+                if (type == NULL || !confignode_hastype(type, CT_BAREVAL)) {
+                    fprintf(stderr, "Invalid type in Attribute in line %hu\n", c->line);
+                    return CKR_DATA_INVALID;
+                }
+                
+                // print attribute by type
+                unsigned int hex = confignode_to_intval(hex_string)->value;
+                if (strcmp((confignode_to_bareval(type)->value), "CK_BBOOL") == 0) {
+                    // build template
+                    CK_BBOOL a_bool;
+                    CK_ULONG bs = sizeof(CK_BBOOL);
+                    CK_ATTRIBUTE temp[] = { {hex, &a_bool, bs} };
+
+                    // get attribute value
+                    rc = funcs->C_GetAttributeValue(session, hkey, temp, 1);
+                    if (rc == CKR_OK) {
+                        if (temp[0].ulValueLen == 0) {
```

Comment:
I cannot find this case in the standard.  If it is not exportable, shouldn't we either get a length of `CK_UNAVAILABLE_INFORMATION`, or even a different return value?

Reply:
Done